File: ACM/dfs_bfs/p11724. 연결 요소의 개수.py
import sys
input = sys.stdin.readline

def dfs(node):
    visited[node] = True
    stack = [node]

    while stack:
        x = stack.pop()
        for i in graph[x]:
            if not visited[i]:
                visited[i] = True
                stack.append(i)

def bfs(node):
    visited[node] = True
    queue = [node]

    while queue:
        x = queue.pop(0)
        for i in graph[x]:
            if not visited[i]:
                visited[i] = True
                queue.append(i)

N, M = map(int ,input().split())

visited = [False] * (N+1)
graph = [[] for _ in range(N+1)]

for i in range(M):
    start, end = map(int, input().split())
    graph[start] += [end]
    graph[end] += [start]

cnt = 0
for i in range(1, N+1):
    if not visited[i]:
        cnt += 1
        # dfs(i)
        bfs(i)

print(cnt)

File: ACM/dfs_bfs/p11724. 연결 요소의 개수.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Main {
	static ArrayList<Integer>[] graph;
	static boolean[] visited;
	
	static void bfs(int node) {
		visited[node] = true;
		Queue<Integer> q = new LinkedList<Integer>();
		q.add(node);
		
		while (!q.isEmpty()) {
			int x = q.poll();
			for (int i : graph[x]) {
				if (!visited[i]) {
					visited[i] = true;
					q.add(i);
				}
			}
		}
	}
	
	static void dfs(int node) {
		if (visited[node]) return;
		
		visited[node] = true;
		for (int i : graph[node]) {
			if (!visited[i])
				dfs(i);
		}
	}
	
    public static void main(String[] args) {
 		Scanner sc = new Scanner(System.in);
 		int N = sc.nextInt();
 		int M = sc.nextInt();
 		
 		graph = new ArrayList[N+1];
 		visited = new boolean[N+1];
 		
 		for (int i=1; i < N+1; i++) {
 			graph[i] = new ArrayList<Integer>();
 		}
 		
 		for (int i=0; i < M; i++) {
 			int start = sc.nextInt();
 			int end = sc.nextInt();
 			
 			graph[start].add(end);
 			graph[end].add(start);
 		}
 		
 		int cnt = 0;
 		for (int i=1; i < N+1; i++) {
 			if (!visited[i]) {
 				cnt++;
// 				dfs(i);
 				bfs(i);
 			}
 		}
 		
 		System.out.println(cnt);
    }
}